087-external-certificate-manager.md

```diff
@@ -0,0 +1,241 @@
+
+Similar to today, Strimzi will also add both these annotations to the pod mounting the Secret.
+
+#### Tracking changes to cluster end-entity certificates
+
+Cert-manager will be responsible for renewing all end-entity certificates.
+When a certificate is renewed cert-manager will update the related Secret.
+During a reconciliation Strimzi will check the hash of the certificate stored in cert-manager Secrets to see if an update is needed.
+
+When cert-manager renews an end-entity certificate Strimzi will copy over the new certificate into its own Secret.
+Strimzi will update the `strimzi.io/server-cert-hash` annotation to match the new certificate.
+Strimzi will update the `strimzi.io/cluster-ca-cert-generation` annotation to match the annotation on the `<CLUSTER_NAME>-cluster-ca-cert` Secret.
+
+#### Handling CA key replacements
+
+During a CA key replacement the user must pause Strimzi reconciliation until the following are complete:
+* A new CA key has been generated
+* A new CA public certificate has been issued
+* The Secret provided to Strimzi containing the CA public certificate has been updated
+* Cert-manager has issued new end-entity certificates using the new CA public certificate
+
+Once these are complete Strimzi can be resumed.
+During the next reconciliation Strimzi will:
+1. Copy over the new CA public cert (keeping the old one as described previously) and update the hash and generation annotations.
+2. Roll the Kafka pods to trust the new CA cert, incrementing the key generation on the pods.
+3. Copy over the new Kafka certificates, updating the cert generation annotations on the Secrets.
+4. Roll the Kafka pods to use the new Kafka certificates, updating the cert generation annotations on the Pods.
+5. Then on the next reconciliation, since the Kafka pods now have correct cert and key generation, copy over the new operator certificate.
+6. Since the Kafka pods now have correct cert generation, remove the old CA public cert from the `<CLUSTER_NAME>-cluster-ca-cert` Secret
+
+I have created diagrams for the existing flow and the flow when using cert-manager for [certificate renewals](./images/087-cert-renewals.png) and the flow for when cert-manager issues [new end-entity certificates](./images/087-new-ee-certs.png).
+
+### User Operator/Clients CA
+TODO
+
+## Affected/not affected projects
+
+This affects the Cluster Operator and User Operator.
+
+## Compatibility
+
+This feature will be optional and disabled by default.
+
+### Migrating to this feature
+
+To start using this feature in an existing Kafka cluster the user must:
+1. Install cert-manager and create an `Issuer`.
+2. Create Secrets to store the Cluster CA and/or Clients CA public certs
+3. Update the `Kafka` resource to have `clusterCa.type` and/or `clientsCa.type` set to `cert-manager.io`, and `certManagerIssuerRef` and `publicCert` configured.
```

Comment:
They should also set `generateCertificateAuthority: false` as this is set to true by default, or we can simply ignore (see my previous comment).